pkg/acs/calacs/acsccd/blev_funcs_postsm4.c

```diff
@@ -36,152 +39,218 @@ static int make_amp_array(const int arr_rows, const int arr_cols, SingleGroup *
+      } else {
+         if (i < 2) {
+            make_amp_array(arr_rows, arr_cols, sg[1], ampInUse, ampdata);
+         } else {
+            make_amp_array(arr_rows, arr_cols, sg[0], ampInUse, ampdata);
+         }
       }
-    }
+       
+      /* calculate "magic square mean" - this works for any single amp */
+      double sum = 0.0;
+      int num = 0;
+      double magic_square_mean = 0.0;
     
-    magic_square_mean = sum / (double) num;
+      {unsigned int j, k;
+      for (j = 2057; j <= 2066; j++) {
+          for (k = 13; k <= 22; k++) {
+              sum += ampdata[arr_cols*j + k];
+              num++;
+          }
+      }}
     
-    /* make amp + gap array */
-    for (j = 0; j < arr_rows; j++) {
-      for (k = 0; k < (arr_cols + ngap_pix); k++) {
-        if (k < arr_cols) {
-          ampdata_gap[(arr_cols + ngap_pix)*j + k] = ampdata[arr_cols*j + k];
-        } else {
-          ampdata_gap[(arr_cols + ngap_pix)*j + k] = magic_square_mean;
+      magic_square_mean = sum / (double) num;
+    
+      /* make amp + gap array */
+      {unsigned int j, k;
+      for (j = 0; j < arr_rows; j++) {
+          for (k = 0; k < (arr_cols + ngap_pix); k++) {
+              if (k < arr_cols) {
+                 ampdata_gap[(arr_cols + ngap_pix)*j + k] = ampdata[arr_cols*j + k];
+              } else {
+                 ampdata_gap[(arr_cols + ngap_pix)*j + k] = magic_square_mean;
+              }
         }
-      }
-    }
+      }}
 
-    /* calculate true DC bias levels */
-    factor = 1.0 - exp(-1.0 / (time_const[i] * serial_freq));
+      /* factor combining time constant and clocking frequency */
+      double factor = 1.0 - exp(-1.0 / (time_const[ampInUse] * serial_freq));
     
-    dc_bias_levels[0] = magic_square_mean * dc_ratio[i];
+      /* calculate true DC bias levels */
+      dc_bias_levels[0] = magic_square_mean * dc_ratio[ampInUse];
     
-    for (j = 1; j < nquad_pix + 1; j++) {
-      dc_bias_levels[j] = ampdata_gap[j-1] * factor * dc_ratio[i] + 
+      {unsigned int j;
+      for (j = 1; j < nquad_pix + 1; j++) {
+          dc_bias_levels[j] = ampdata_gap[j-1] * factor * dc_ratio[ampInUse] + 
                             (1.0 - factor) * dc_bias_levels[j-1];
-    }
+      }}
     
-    /* calculate correction to data */
-    for (j = 0; j < nquad_pix; j++) {
-      ampdata_gap[j] = (ampdata_gap[j] - dsi_sens[i] * dc_bias_levels[j+1]) -
+      /* calculate correction to data */
+      {unsigned int j;
+      for (j = 0; j < nquad_pix; j++) {
+          ampdata_gap[j] = (ampdata_gap[j] - dsi_sens[ampInUse] * dc_bias_levels[j+1]) -
                           (10./22.) * (dc_bias_levels[j+1] - dc_bias_levels[j]);
-    }
+      }}
     
-    /* copy corrected data back to ampdata */
-    for (j = 0; j < arr_rows; j++) {
-      for (k = 0; k < arr_cols; k++) {
-        ampdata[arr_cols*j + k] = ampdata_gap[(arr_cols + ngap_pix)*j + k];
-      }
-    }
+      /* copy corrected data back to ampdata */
```

Comment:
Is it not possible to combine this op within the loop above (correction comp) and assign the correction directly to  ``ampdata``? This would be more efficient.